signature/signature-v4.go

```diff
@@ -134,9 +140,17 @@ func V4SignVerify(r *http.Request) ErrorCode {
 	// Copy request.
 	req := *r
 	hashedPayload := getContentSha256Cksum(r)
+	queryf := req.URL.Query()
 
 	// Save authorization header.
 	v4Auth := req.Header.Get(headerAuth)
+	if v4Auth == "" {
+		// try query based
+		v4Auth = fmt.Sprintf("%s Credential=%s,SignedHeaders=%s, Signature=%s", queryf.Get(amzAlgorithm), queryf.Get(amzCredential), queryf.Get(amzSignedHeaders), queryf.Get(amzSignature))
+		if v4Auth == "" {
```

Comment:
This will never be true. What is this test for?

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
Please have a look as a reference. When the client generates a web url for easier accessing on web browsers, the authorization headers are passed in query-mode instead of headers. In this case, the v4Auth will be "" and we can get the  auth params from the url query.

Reply:
This code change is for support for "pre-signed URLs", a feature quite commonly used in S3.

Example:

```javascript
import AWS from "aws-sdk";

const s3 = new AWS.S3({
    endpoint: 'https://127.0.0.1',
    accessKeyId: 'aaaa',
    secretAccessKey: 'ssss',
    s3ForcePathStyle: true,
    signatureVersion: 'v4',
    sslEnabled: true,
});


function generateDownloadUrl(bucket, key) {
    const params = {
        Bucket: bucket,
        Key: key,
        Expires: 3600,
    };

    s3.getSignedUrl('getObject', params, (err, url) => {
        if (err) {
            console.error(err);
        } else {
            console.log(url);
        }
    });
}

generateDownloadUrl('bucket1', 'a.txt');
```

It produces URLs like

> https://127.0.0.1/bucket1/a.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=aaaa%2F20240710%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240710T053620Z&X-Amz-Expires=3600&X-Amz-Signature=623b5d21aa2698af14d046d40008a72ef4ba3dfed9bb90d412e6dba4a6ec8706&X-Amz-SignedHeaders=host

This URL can be used anywhere - send a GET request and it will give you the file.

Reply:
I think we are talking at cross purposes! I agree with what you said above but it wasn't my question.

My specific query is this:

In this line we set `v4Auth` to something

```go
v4Auth = fmt.Sprintf("%s Credential=%s, SignedHeaders=%s, Signature=%s", queryf.Get(amzAlgorithm), queryf.Get(amzCredential), queryf.Get(amzSignedHeaders), queryf.Get(amzSignature))
```

This can never be `""`

And in the line after we check it against `""` which can never be true surely?

```go
if v4Auth == "" {
```

Which is indicative of some kind of logic error.

Reply:
Uh sorry, it's my fault, I'm going to correct it.

Reply:
Thanks for correction. I've corrected this logical error, pls have a look at the new force-pushed code.